snuba_sdk/utils.py

```diff
@@ -0,0 +1,138 @@
+from datetime import datetime
+from typing import Any, List, Mapping, Sequence, Union
+
+from snuba_sdk.conditions import Condition, Op
+from snuba_sdk.entity import Entity
+from snuba_sdk.expressions import (
+    Column,
+    Direction,
+    Function,
+    LimitBy,
+    OrderBy,
+)
+from snuba_sdk.query import Query
+from snuba_sdk.query_visitors import InvalidQuery
+
+
+def parse_datetime(date_str: str) -> datetime:
+    return datetime.strptime(date_str, "%Y-%m-%dT%H:%M:%S.%f")
+
+
+def json_to_snql(body: Mapping[str, Any], entity: str) -> Query:
+    def to_exp(value: Union[str, List[Any]]) -> Union[Column, Function]:
+        if not isinstance(value, list):
+            return Column(value)
+
+        children = []
+        if isinstance(value[1], list):
+            children = list(map(to_exp, value[1]))
+        elif value[1]:
+            children = [to_exp(value[1])]
+
+        alias = value[2] if len(value) > 2 else None
+        return Function(value[0], children, alias)
+
+    def to_scalar(value: Any) -> Any:
+        if isinstance(value, list):
+            return list(map(to_scalar, value))
+        elif isinstance(value, tuple):
+            return tuple(map(to_scalar, value))
+
+        if isinstance(value, str):
+            try:
+                date_scalar = parse_datetime(value)
+                return date_scalar
+            except ValueError:
+                return value
+
+        return value
+
+    dataset = body.get("dataset", "")
+    query = Query(dataset, Entity(entity))
+
+    selected_columns = list(map(to_exp, body.get("selected_columns", [])))
+    for a in body.get("aggregations", []):
+        if a[0].endswith(")") and not a[1]:
+            selected_columns.append(Function(a[0].strip("()"), [], a[2]))
```

Comment:
I believe the visibility team is still using infix expressions in aggregations. Mind adding a check for such expressions (checking for the presence of parentheses after you removed these) and fail explicitly?

Reply:
There's no need for an explicit exception, this will throw an InvalidExpression exception since "(" are not allowed in function names. However should I try to unravel the function? I'm just worried that if this function doesn't translate any JSON snuba query to SnQL, then the migration to SnQL will stall while we wait for the product teams to start using the SDK properly.

Reply:
> There's no need for an explicit exception, this will throw an InvalidExpression exception since "(" are not allowed in function names.

That is what I meant by adding an explicit expression. Explain in the exception that the behavior is not supported. Otherwise people will keep trying to port those queries and complain when they get this cryptic InvalidExpression message.

> However should I try to unravel the function
If you are talking about parsing the function, I would discourage that as you would have to port the parser and its grammar here. But that's up to you.